### PR TITLE
Allow peribolos to manage repo permissions for teams

### PR DESCRIPTION
**Pull request by stevekuznetsov**
With this patch, the configuration for Peribolos/org management can now
be used to enforce a specific set of repository permissions for the
groups that it manages.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @fejta @cjwagner

### CODE REVIEW

File: prow/config/org/org.go [hunk 1]
```diff
@@ -60,44 +62,13 @@ type Team struct {
 	Children    map[string]Team `json:"teams,omitempty"`
 
 	Previously []string `json:"previously,omitempty"`
-}
-
-// RepoPermissionLevel is admin, write, read or none.
-//
-// See https://developer.github.com/v3/repos/collaborators/#review-a-users-permission-level
-type RepoPermissionLevel string
```
Comment (stevekuznetsov) [inline prow/config/org/org.go]: Moving this was necessary to break an import cycle

**Comment by stevekuznetsov** [comment]
Added more unit tests.

File: prow/cmd/peribolos/README.md [hunk 2]
```diff
@@ -46,6 +46,9 @@ orgs:
         - anne
         maintainers:
         - jane
+        repos: # Ensure the team has the following permissions levels on repos in the org
+          some-repo: admin
```
Comment (fejta) [inline prow/cmd/peribolos/README.md]: Once we have repo management I feel like this would be better to scope under the repo.

There's no guarantee that a `some-repo` exists (although I guess one could also argue that there's no guarantee that `some-team` exists)

**Comment by stevekuznetsov** [comment]
If you `--dump` first it should have no effect on GitHub state, so I thought gating it would not be necessary. WDYT?

File: prow/cmd/peribolos/README.md (hunk 2 above)
Comment (stevekuznetsov) [reply prow/cmd/peribolos/README.md]: I thought about this, too, but:

1. GitHub's API works through the team, so I figured it's best if we mirror it for code complexity
2. Running on repo, not on team, removes the nice property for dumps above and will make honoring the flag to ignore teams harder
3. Running per repo also adds a substantial API burden to list repo state through teams

**Comment by fejta** [comment]
Most of the other features are flag-gated: 

https://github.com/kubernetes/test-infra/blob/1e7bc35c0a7680f8d3879292c5704f869453b757/prow/cmd/peribolos/main.go#L85-L88

**Comment by stevekuznetsov** [comment]
Ahh, missed that -- added a gating flag.

**Comment by stevekuznetsov** [comment]
@fejta updated the PR to rebase on #12154  and added a debugging commit which we can either drop or run in a different PR, whichever you prefer

**Comment by stevekuznetsov** [comment]
Fixed the flag test, made `repos` `omitempty` for `--dump`. @fejta @cblecker @nikhita this is ready for review :)

File: prow/github/client.go [hunk 3]
```diff
@@ -2186,6 +2185,78 @@ func (c *Client) ListTeamMembers(id int, role string) ([]TeamMember, error) {
 	return teamMembers, nil
 }
 
+// ListTeamRepos gets a list of team repos for the given team id
+//
+// https://developer.github.com/v3/teams/#list-team-repos
+func (c *Client) ListTeamRepos(id int) ([]Repo, error) {
```
Comment (fejta) [inline prow/github/client.go]: I'd be inclined to put the client changes in their own commit

Comment (stevekuznetsov) [reply prow/github/client.go]: done
